fn/goroutine_manager.go

```diff
@@ -0,0 +1,66 @@
+package fn
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+// ErrStopping is returned when trying to add a new goroutine while stopping.
+var ErrStopping = errors.New("can not add goroutine, stopping")
+
+// GoroutineManager is used to launch goroutines until context expires or the
+// manager is stopped. The Stop method blocks until all started goroutines stop.
+type GoroutineManager struct {
+	wg     sync.WaitGroup
+	mu     sync.Mutex
+	ctx    context.Context
+	cancel func()
+}
+
+// NewGoroutineManager constructs and returns a new instance of
+// GoroutineManager.
+func NewGoroutineManager(ctx context.Context) *GoroutineManager {
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &GoroutineManager{
+		ctx:    ctx,
+		cancel: cancel,
+	}
+}
+
+// Go starts a new goroutine if the manager is not stopping.
+func (g *GoroutineManager) Go(f func(ctx context.Context)) error {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	if g.ctx.Err() != nil {
+		return ErrStopping
+	}
+
+	g.wg.Add(1)
+	go func() {
+		defer g.wg.Done()
+		f(g.ctx)
+	}()
+
+	return nil
+}
+
+// Stop prevents new goroutines from being added and waits for all running
+// goroutines to finish.
+func (g *GoroutineManager) Stop() {
+	g.mu.Lock()
+	g.cancel()
+	g.mu.Unlock()
+
+	// Wait for all goroutines to finish.
+	g.wg.Wait()
+}
+
+// Context returns internal context of the GoroutineManager which will expire
+// when either the context passed to NewGoroutineManager expires or when Stop
+// is called.
+func (g *GoroutineManager) Context() context.Context {
+	return g.ctx
+}
```

Comment:
I strongly recommend against leaking the context from the manager.

Reply:
Could you elaborate on this, because the ErrGroup implementation does even return the context to be used in the spawning goroutines?

Reply:
The point of this abstraction is to encapsulate a set of goroutines and have it manage the lifetime of the context. If you leak the context outside the GoroutineManager, then it no longer can manage the lifetime of the context. It destroys your ability to reason about the abstraction.

Reply:
ahh ok makes sense, and we never know when the context might be used as a parent for another group. Thinking about it twice yes there is really no need in leaking this.

Reply:
I added this method to be able to wait until GoroutineManager starts stopping. This context replaces `quit` channel in the original code, see an example here: https://github.com/lightningnetwork/lnd/pull/9140
Some of the code waiting for `quit` does not run inside a goroutine launched by GoroutineManager, so it doesn't have access to ctx argument passed by `Go` method to a function.
What approach would you prefer to wait until GoroutineManager starts stopping?

Reply:
I think we should still keep the context hidden, because in your provided example you have all the information you need when calling the Stop() method, which means your program should decided upon the Stop() finishing rather than listening to the internal Manger context. So we should probably still keep the quit channel and either close it before calling the goroutine Manager or after ?

Reply:
If you need a method that waits on the context stopping then add the wait method. Do not leak the context. The purpose of this abstraction is to encapsulate the context. If that isn't its purpose then I don't know what is. If you want to encapsulate the context and give it a clean and safe API then you undermine that goal by leaking it outside of the struct.

Reply:
The code using this manager relies on a channel and `select` to wait not only on stopping, but on some other things, e.g. (from `switch.go`):

```go
select {
	case <-linkQuit:
		return nil
	case <-s.quit:
		return nil
	default:
		...
}
```

If I replace s.quit with `GoroutineManager.Wait()` call, it would be hard to plug this into existing code using `select` mostly and the scope of the change would be broader, which I want to avoid. I want a drop-in replacement for a context/channel and a WaitGroup in places where we use them for goroutines tracking.

@ziggie1984  If I keep `quit` channel and close it when calling Stop, then it would not be closed if the context passed to `NewGoroutineManager` is cancelled. It would be an additional logic in switch.go to keep them in sync.

> The purpose of this abstraction is to encapsulate the context. If that isn't its purpose then I don't know what is.

The purpose of this abstraction is to have an easy way of launching goroutines until shutdown and wait for them to complete. It turned out to be challenging to do it properly with just `sync` package (`sync.WaitGroup` + quit channel), that is why this abstraction is needed to prevent the use of a WaitGroup resulting in a race condition.

What do you think about replacing `Context()` method with `Done()` method returning just a "done" receive-only channel of the context?

Reply:
> What do you think about replacing Context() method with Done() method returning just a "done" receive-only channel of the context?

This is vastly superior because it solves your actual problem without exposing the entire context api to consumers who may break invariants that this abstraction assumes.

Reply:
I replaced `Context()` method with `Done()` method returning `<-chan struct{}`.